### PR TITLE
ZENKO-3165 implementing sphinx copybutton plugin

### PR DESCRIPTION
**Pull request by tcarmet**

### CODE REVIEW

File: docs/docsource/conf.py
Comment (miniscruff) [inline docs/docsource/conf.py]: this line is out of alignment...

File: docs/docsource/requirements.in
Comment (miniscruff) [inline docs/docsource/requirements.in:10]: is this not on PyPI? It definitely should be... 😏

Comment (wabernat) [reply docs/docsource/requirements.in:10]: @tcarmet: MeanINGful...
@miniscruff: Teachable moment here: up until now, I thought PyPi was a command that pointed to some mystic cloud. There is a repository as well?

Comment (miniscruff) [reply docs/docsource/requirements.in:10]: Yup you upload your packages to this index so that pip can download them.

**Comment by wabernat** [comment]
I've tested it, and it's ALMOST there. It builds and it functions correctly. Hooooowwwwevver... there is a minor cosmetic issue: the little clipboard icon is not appearing in the right margin (I assume this is a missing asset). I see a degree sign (°) instead: 

image
vs.
image

File: docs/docsource/requirements.in
Comment (wabernat) [reply docs/docsource/requirements.in:10]: That takes much of the mystique out of that cloud.

**Comment by miniscruff** [comment]
@wabernatScality I think it's just tiny, I zoomed in a bit

**Comment by wabernat** [comment]
Right you are... 

This is bearing the whiff of css.

**Comment by wabernat** [comment]
Manually editing the CSS `height` attribute in the output web page to `14.4pt` achieves the desired appearance:

image

**Comment by wabernat** [comment]
Okay... so for some reason, when we do this in Zenko, the `a.cpybtn>image>height` attribute is not written to the css, and it's defaulting to "dinky".

**Comment by wabernat** [comment]
@nootal If we haven't figured this out by the time you get this (don't wreck your weekend), there's an issue that looks like it may be coming from your sphinx_copybutton fork (not sure, but it's not translating CSS to Zenko the same as it was for docs-ring-s3c)

File: docs/docsource/requirements.in
Comment (tcarmet) [reply docs/docsource/requirements.in:10]: Oups, thanks @wabernatScality gonna fix that. Also worth noting that I have no plans in merging this as it is. (tried to open the PR as draft, but I failed :laughing: ) 

@miniscruff so `sphinx-copybutton`  is a real sphinx plugin and it's available on pypi. However I discovered that a fork has been made internally with some patches that I would even advice to make a PR upstream. That not being the case what's advisable to move forward here? My thinking is, to at least try to push some of the changes on the upstream repository. But there might be some of the patches that would not be "accepted" by the community, so we may have to "live" with the fork or find another way. I don't feel like uploading a fork of a project to pypi, but I'm curious to know if, even if it's a fork we should upload to pypi.

File: docs/docsource/conf.py
Comment (tcarmet) [reply docs/docsource/conf.py]: hmm, funny, it wasn't on my editor. Probably a `tab` hidden :detective:

Comment (tcarmet) [reply docs/docsource/conf.py]: done

**Comment by tcarmet** [comment]
@wabernatScality I would advice us to give a shot with the opensource `sphinx-copybutton` extensions (without scality patches) to see if our patches are useful for Zenko documentation, and if we can live without the list in 
```python
copybutton_prompt_text = [
    '$ ',
    '# ',
    '[root@scality] # ',
    '[root@scality]# ',
    '[doc@scality]$ ',
]
```

Maybe we can modify the doc so that `copybutton_prompt_text = '$ '` would be enough. 

What's your thought on this?

File: docs/docsource/requirements.in
Comment (miniscruff) [reply docs/docsource/requirements.in:10]: Changes that reflect a global use case should be pushed up stream for sure, anything that will not can go into our fork. Uploading a fork to pypi is fine, just prefix it with scality and provide a link to the main base. Making sure we keep in line as much as possible. Also, making sure the license allows. Imo.

**Comment by wabernat** [comment]
The reason @nootal implemented local changes/forked the plugin was because the unmodified plugin's performance is inadequate. Without the forked code, the plugin only recognized one prompt (I set it to '$ ') and could not recognize any other prompt. It claimed to be able to handle regexp, but I could not make it work. 

My expectation with this and all other mods to the default sphinx tool set it to _unify_ build tools, not to make more inter-product exceptions. In other words, if I could get both docs-ring-s3c and Zenko to use the exact same settings and fill in the details (the list of prompts in conf.py) using variables, or even a reference to a single set published somewhere containing _all_ prompts in all documentation, that's the direction I would head. 

If I understand you correctly, what you're proposing runs counter to that, and I would oppose it. I also agree that it would be worthwhile for us to push our mods to the main branch for the plugin, assuming the code is generalizable enough to be useful to other users. If both you and I can't get the plugin to work correctly with the mods, this is useful information, and may help us improve the plugin to the point that it can be published with pride.

**Comment by wabernat** [comment]
I removed custom.css and classic.css because I was looking for CSS defs I could modify, and I noticed that these files were absent in docs-ring-s3c, which builds just fine. I renamed them and Zenko built fine, so I deleted them. I believe most of the template heavy lifting goes on at https://github.com/scality/sphinx-tools. Why we get this ever-so-slight difference between Zenko and docs-ring-s3c remains a mystery, but there's no harm in jettisoning this valueless cargo along the way.

**Comment by miniscruff** [comment]
> I removed custom.css and classic.css because I was looking for CSS defs I could modify, and I noticed that these files were absent in docs-ring-s3c, which builds just fine. I renamed them and Zenko built fine, so I deleted them. I believe most of the template heavy lifting goes on at https://github.com/scality/sphinx-tools. Why we get this ever-so-slight difference between Zenko and docs-ring-s3c remains a mystery, but there's no harm in jettisoning this valueless cargo along the way.

:+1: for giving CSS the boot

**Review by wabernat — COMMENTED** [review]
I will approve the moment we get the clipboard icon size issue resolved. I can't fix this. If you can't fix it either, say "uncle" and we'll see if @nootal has the skill.

**Comment by wabernat** [comment]
Apparently, the tiny icon issue is due to my local instance of Zenko having become corrupted. Downloading a fresh copy seems to work fine.

**Comment by miniscruff** [comment]
Once everything is building right we should squash the commits, we are deploying alpha 3 very soon though so I would wait on this just to keep the machines happy.

**Comment by wabernatScality** [comment]
.

**Comment by wabernat** [comment]
I may have made a comment in the wrong place. I cloned a fresh Zenko repo and the "tiny icon" issue disappeared. I conclude that this was a case of local corruption on my working Zenko build. If I had a place to do so, I would approve this PR. I exhort you to do the same!

**Review by wabernat — APPROVED** [review]

**Comment by wabernat** [comment]
@tcarmet I think this is just waiting for your approval

**Comment by tcarmet** [comment]
I need to fix this part first https://github.com/scality/Zenko/pull/1122/files#diff-462ae7efd4e820a1dfedd50432320f7d9e51bd1861b9f0d80a5d86fd03fef811R9 or at least have a tagged version on that repo instead of pulling a branch name, before I can merge this PR.

**Review by wabernat — APPROVED** [review]

**Comment by wabernat** [comment]
This structure is downright confusing! Anyway...I guess I've approved the hell out of this thing....

**Comment by wabernat** [comment]
@tcarmet What's hanging this up?

**Review by lavigne958 — APPROVED** [review]

**Pull request closed without merging**
